File: model/control_allocator.m
function [output, B] = control_allocator(input, params, mode)
% control_allocator: Quadrotor control allocation
%
% Usage:
%   [omega_sq, B] = control_allocator(cmd_vec, params, 'inverse')
%   [cmd_vec, B]  = control_allocator(omega_sq, params, 'forward')
%
% Inputs:
%   input  - 4x1 vector
%            'inverse': [T; τx; τy; τz] → omega_sq
%            'forward': [ω0²; ω1²; ω2²; ω3²] → [T; τx; τy; τz]
%   params - parameter struct
%   mode   - 'inverse' (default) or 'forward'
%
% Outputs:
%   output - 4x1 vector
%   B      - 4x4 allocation matrix

if nargin < 3
    mode = 'inverse';
end

%% Parameters
k_T = params.drone.motor.k_T;   % Thrust coeff [N/(rad/s)^2]
k_M = params.drone.motor.k_M;   % Moment coeff [m]: τ = k_M * T

%% Lever Arms (absolute values)
l_x  = 0.13;   % [m] longitudinal arm
l_y1 = 0.22;   % [m] lateral arm (rotor 0, 2)
l_y2 = 0.20;   % [m] lateral arm (rotor 1, 3)

%% Allocation Matrix B
%  [T; τx; τy; τz] = B * [ω1²; ω2²; ω3²; ω4²]
%
%  Coordinate: X-Forward, Y-Right, Z-Down (NED)
%
%  Rotor layout (top view):
%
%           x (Front)
%           ^
%           |
%     1(CCW)|  3(CW)
%        \  |  /
%         \ | /
%          -+----> y (Right)
%         / | \
%        /  |  \
%     4(CW) |  2(CCW)
%           |
%        (Back)
%
%  Rotor 1: Front-Left,  CCW, (+l_x, -l_y1)
%  Rotor 2: Back-Right,  CCW, (-l_x, +l_y2)
%  Rotor 3: Front-Right, CW,  (+l_x, +l_y1)
%  Rotor 4: Back-Left,   CW,  (-l_x, -l_y2)
%
%  Torque convention (τ = r × F, F = [0;0;-T]):
%    τx = -y*T  (left rotor → +roll, right rotor → -roll)
%    τy = +x*T  (front rotor → +pitch, back rotor → -pitch)
%    τz = reaction torque (CCW rotor → +yaw, CW rotor → -yaw)

%            ω1²        ω2²        ω3²        ω4²
%           ─────      ─────      ─────      ─────
B = [
% T:    all thrusts sum up
        k_T,        k_T,        k_T,        k_T;

% τx:   Left(+)    Right(-)   Right(-)   Left(+)
        k_T*l_y1,  -k_T*l_y2,  -k_T*l_y1,  k_T*l_y2;

% τy:   Front(+)   Back(-)    Front(+)   Back(-)
        k_T*l_x,   -k_T*l_x,   k_T*l_x,  -k_T*l_x;

% τz:   CCW(+)     CCW(+)     CW(-)      CW(-)
        k_T*k_M,    k_T*k_M,  -k_T*k_M,  -k_T*k_M
];

%% Compute output
if strcmp(mode, 'inverse')
    output = B \ input;
else
    output = B * input;
end

end

File: mex/hexa_mex_ref/src/control_allocator.cpp
#include "../include/types.hpp"
#include "../include/params.hpp"
#include <cmath>
#include <cstring>

// Hexarotor control allocation
// Motor layout (top view, looking down Z+):
//           X+ (Front)
//            ↑
//       M2      M1
//      (CW)   (CCW)
//         \    /
//  M3 -----+-----→ Y+ (Right)
// (CCW)    |    M6 (CW)
//         /    \
//      M4      M5
//     (CW)    (CCW)
//
// Allocation Matrix B (4x6):
//   [T; τx; τy; τz] = B * [ω1²; ω2²; ω3²; ω4²; ω5²; ω6²]

void control_allocator_inverse(const float* cmd_vec, float* omega_sq, const Params& params) {
    // cmd_vec[0] = T (total thrust)
    // cmd_vec[1] = tau_x (roll torque)
    // cmd_vec[2] = tau_y (pitch torque)
    // cmd_vec[3] = tau_z (yaw torque)

    // Suppress unused parameter warning
    (void)params;

    float T = cmd_vec[0];
    float tau_x = cmd_vec[1];
    float tau_y = cmd_vec[2];
    float tau_z = cmd_vec[3];

    // Allocation matrix B (4x6):
    // Row 0: [ k,  k,  k,  k,  k,  k ]
    // Row 1: [-kL/2, kL/2, kL, kL/2, -kL/2, -kL]
    // Row 2: [kL*s3, kL*s3, 0, -kL*s3, -kL*s3, 0]
    // Row 3: [b, -b, b, -b, -b, b]  ← NOTE: M6 is CW (last -b changed to +b)

    // Solve B * omega_sq = cmd_vec using least-squares (pinv(B) * cmd_vec)
    // Precomputed pseudo-inverse for hexarotor (from MATLAB control_allocator.m)
    //
    // For hexarotor, B is 4x6, so pinv(B) is 6x4
    // Computed using: pinv_B = B' * inv(B * B')

    // Simplified analytical solution (from MATLAB pinv computation):
    // This is derived from the specific geometry of hexarotor

    // Use exact pseudo-inverse values computed from actual B matrix with physical units
    // pinv(B) where B has k_T, k_T*L, k_T*L*sqrt(3)/2, k_M coefficients
    const float pinv[6][4] = {
        {1.2919896641e+03f, -1.3458225668e+03f,  2.3310330636e+03f,  7.1530758226e+03f},  // M1
        {1.2919896641e+03f,  1.3458225668e+03f,  2.3310330636e+03f, -7.1530758226e+03f},  // M2
        {1.2919896641e+03f,  2.6916451335e+03f,  0.0f,               7.1530758226e+03f},  // M3
        {1.2919896641e+03f,  1.3458225668e+03f, -2.3310330636e+03f, -7.1530758226e+03f},  // M4
        {1.2919896641e+03f, -1.3458225668e+03f, -2.3310330636e+03f,  7.1530758226e+03f},  // M5
        {1.2919896641e+03f, -2.6916451335e+03f,  0.0f,              -7.1530758226e+03f}   // M6
    };

    // Apply pseudo-inverse directly to command vector (no additional scaling needed)
    for (int i = 0; i < 6; i++) {
        omega_sq[i] = pinv[i][0] * T +
                      pinv[i][1] * tau_x +
                      pinv[i][2] * tau_y +
                      pinv[i][3] * tau_z;
    }
}

void control_allocator_forward(const float* omega_sq, float* cmd_vec, const Params& params) {
    // Forward: [T; τx; τy; τz] = B * omega_sq

    float k = params.drone.k_T;
    float b = k * params.drone.k_M;
    float L = params.drone.L;
    float s3 = sqrtf(3.0f) / 2.0f;

    float kL = k * L;
    float kL_s3 = kL * s3;

    // Total thrust
    cmd_vec[0] = k * (omega_sq[0] + omega_sq[1] + omega_sq[2] + omega_sq[3] + omega_sq[4] + omega_sq[5]);

    // Roll torque (tau_x)
    cmd_vec[1] = -kL/2.0f * omega_sq[0] + kL/2.0f * omega_sq[1] + kL * omega_sq[2]
                 + kL/2.0f * omega_sq[3] - kL/2.0f * omega_sq[4] - kL * omega_sq[5];

    // Pitch torque (tau_y)
    cmd_vec[2] = kL_s3 * omega_sq[0] + kL_s3 * omega_sq[1]
                 - kL_s3 * omega_sq[3] - kL_s3 * omega_sq[4];

    // Yaw torque (tau_z): CCW motors (+b), CW motors (-b)
    // M1(CCW), M2(CW), M3(CCW), M4(CW), M5(CCW), M6(CW)
    cmd_vec[3] = b * omega_sq[0] - b * omega_sq[1] + b * omega_sq[2]
                 - b * omega_sq[3] + b * omega_sq[4] - b * omega_sq[5];
}
